src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,760 @@
+            // into whatever start is pointing to. O(n)
+            while (start <= end)
+            {
+                if (items[start].Key != null)
+                {
+                    start++;
+                }
+                else if (items[end].Key != null)
+                {
+                    // Swap this item into start and advance
+                    items[start] = items[end];
+                    items[end] = default;
+                    start++;
+                    end--;
+                }
+                else
+                {
+                    // Both null, we need to hold on 'start' since we
+                    // still need to fill it with something.
+                    end--;
+                }
+            }
+
+            return new SmallCapacityDictionary<TKey, TValue>()
+            {
+                _arrayStorage = items!,
+                _count = start,
+            };
+        }
+
+        /// <summary>
+        /// Creates an empty <see cref="SmallCapacityDictionary{TKey, TValue}"/>.
+        /// </summary>
+        public SmallCapacityDictionary()
+            : this(0, EqualityComparer<TKey>.Default)
+        {
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SmallCapacityDictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <param name="comparer">Equality comparison.</param>
+        public SmallCapacityDictionary(IEqualityComparer<TKey> comparer)
+            : this(0, comparer)
+        {
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SmallCapacityDictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <param name="capacity">Initial capacity.</param>
+        public SmallCapacityDictionary(int capacity)
+            : this(capacity, EqualityComparer<TKey>.Default)
+        {
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SmallCapacityDictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <param name="capacity">Initial capacity.</param>
+        /// <param name="comparer">Equality comparison.</param>
+        public SmallCapacityDictionary(int capacity, IEqualityComparer<TKey> comparer)
+        {
+            if (comparer is not null && comparer != EqualityComparer<TKey>.Default) // first check for null to avoid forcing default comparer instantiation unnecessarily
+            {
+                _comparer = comparer;
+            }
+            else
+            {
+                _comparer = EqualityComparer<TKey>.Default;
+            }
+
+            if (capacity == 0)
+            {
+                _arrayStorage = Array.Empty<KeyValuePair<TKey, TValue>>();
+            }
+            else if (capacity <= DefaultArrayThreshold)
+            {
+                _arrayStorage = new KeyValuePair<TKey, TValue>[capacity];
+            }
+            else
+            {
+                _backup = new Dictionary<TKey, TValue>(capacity: 10);
```

Comment:
```suggestion
                _backup = new Dictionary<TKey, TValue>(capacity);
```
Why 10 when the capacity is given as argument?